provider/ec2/environ.go

```diff
@@ -110,27 +110,30 @@ func (e *environ) Name() string {
 
 // PrepareForBootstrap is part of the Environ interface.
 func (env *environ) PrepareForBootstrap(ctx environs.BootstrapContext) error {
+	callCtx := context.NewCloudCallContext()
+	// Cannot really invalidate a credential here since nothing is bootstrapped yet.
+	callCtx.InvalidateCredentialFunc = func(string) error { return nil }
```

Comment:
Maybe doing nothing should be the default behaviour of the context? Would you get a nil-pointer error if this hadn't been set?

Reply:
No, if InvalidateCredentialFunc is not overwritten as above, it will not get a nil pointer but will tell you that this method has not been implemented instead.

I want to keep it this as our default behavior to catch any strange and obscure code paths that may have been missed in my excessive testing. After all, there is always a tiny, remote chance that there is a path that has not been walked by me. I also believe that getting an explicit "not implemented" yet in the log will help diagnose these obscure cases.

So, I would prefer to keep the default as "not implemented" and, in cases such as this one, to have a silent implementation (i.e. returning nil). In this particular case, there is nothing that we can do with an invalid credential -  we have not yet even stored it on the controller and, hence, cannot mark our copy as invalid.

Reply:
Yeah, that makes sense - I'm happy with needing an explicit do nothing in the rare cases you need that. Having a default that highlights the problem without panicing is good too.